### PR TITLE
Make the Marine DA almost work again

### LINKED ISSUE #393: Remove empty obs spaces
### Description
What we have in `obs_list.yaml` is what we wish to assimilate/simulate but the staged obs will rarely have the entire wish list available. The result is the jedi exe failing.
 
### Solution
There are multiple ways to do this. I'm thinking about culling the obs space based on what was actually staged. Any other suggestions @RussTreadon-NOAA , @CoryMartin-NOAA , @AndrewEichmann-NOAA , ... and others ...

A simple python application:
```
input: 
 - var/hofx/letkf.yaml
 - path to where the obs were staged
output:
- new var.yaml with obs spaces pointing to non-existent files removed
```

### LINKED ISSUE #415: Interpolate the Tref increment to the MOM6 grid
This is all out of my not too reliable memory ... So, there's that.
We wouldn't mind making use of the Tref increment computed as part of the surface analysis (NSST).

Here are the needed steps:
1 - Compute the Tref increment on the Gaussian grid where the ocean mask is defined, zero out where no ocean
2 - Interpolate the increment from Gaussian to tripolar using the pre-computed weights

### LINKED ISSUE #422: Sea ice analysis insertion
See here:
https://github.com/guillaumevernieres/global-workflow/blob/f1f125a84cb1f5ed2fcc7b7466402415b3e0398f/ush/forecast_postdet.sh#L1003

### LINKED ISSUE #423: Add icepack to the GDASApp bundle

### LINKED ISSUE #424: Update the saber block yaml

### LINKED ISSUE #427: Update 3DVAR FGAT yaml

### LINKED ISSUE #428: Consolidate the logging of the B-mat verify task

### LINKED ISSUE #429: Fix bug in pseudo model!
We were passing the background with a 3 hour offset ... oops. The bug is in this script:
```scripts/exgdas_global_marine_analysis_prep.py```

### LINKED ISSUE #430: The run j-job should only run the var.x executable
related to issue #418

### LINKED ISSUE #431: Add seaice diagnostics to the verify j-job

### LINKED ISSUE #433: Update the run_jjob.py application

### LINKED ISSUE #434: Bug in ush/socaincr2mom6.py

### LINKED ISSUE #435: Update the dirac yaml generator application

### LINKED ISSUE #436: Bug in the R2D2 fetching of the PT10M obs

### LINKED ISSUE #437: add a ctest for soca2mom6 increment, include the nsst merging.

### LINKED ISSUE #440: Add the checkpoint j-job to the ctest suite

### LINKED ISSUE #443: Fix the SOCA BUMP initialization

### PR DESCRIPTION
**Pull request by guillaumevernieres**
## Description
This PR is the one before last merge that will allow cycling the marine DA within the global-workflow WCDA system (excluding the minor issue of the recent reorg.). @AndrewEichmann-NOAA is responsible for merging the penultimate work. No pressure.

Most of the work was done by @AndrewEichmann-NOAA and myself.

## Work Done
- Addition of ```icepack``` to the bundle
- Fixed the ```yaml``` 's for the bump initialization
- Increased the domain stack size for the ```soca``` to ```cice6``` change of variable
- Added a "checkpoint" script
- Fixed a major bug in the pseudo model configuration
- Added a python function to cleanup empty obs spaces
- Added a few more figures in the vrfy script
- Added a ctest for the soca/nsst increment to mom6 application
- Fixed bug in ```run_jjob.py``` script 
- Fixed bug in the dirac application
- Fixed bug in the fetching of the PT10M obs

## Issues addressed
- fixes #443 
- fixes #440 
- fixes #437 
- fixes #436 
- fixes #435 
- fixes #434 
- fixes #433 
- fixes #431 
- fixes #430 
- fixes #429 
- fixes #428 
- fixes #427 
- fixes #424 
- fixes #423 
- fixes #422 
- fixes #415 
- fixes #393

### CODE REVIEW

File: test/soca/socaincr2mom6.sh [hunk 1]
```diff
@@ -0,0 +1,20 @@
+#!/bin/bash
+set -ex
+
+srcdir=$1
+
+data="/home/gvernier/wcda-sprint/wrk"
```
Comment (guillaumevernieres) [inline test/soca/socaincr2mom6.sh]: brilliant work once again @guillaumevernieres

Comment (AndrewEichmann-NOAA) [reply test/soca/socaincr2mom6.sh]: We should strive for portability, after all

File: test/soca/gw/CMakeLists.txt [hunk 2]
```diff
@@ -32,8 +32,9 @@ set(jjob_list "JGDAS_GLOBAL_OCEAN_ANALYSIS_PREP"
               "JGDAS_GLOBAL_OCEAN_ANALYSIS_BMAT"
               "JGDAS_GLOBAL_OCEAN_ANALYSIS_BMAT_VRFY"
               "JGDAS_GLOBAL_OCEAN_ANALYSIS_RUN"
-              "JGDAS_GLOBAL_OCEAN_ANALYSIS_POST"
-              "JGDAS_GLOBAL_OCEAN_ANALYSIS_VRFY")
+              "JGDAS_GLOBAL_OCEAN_ANALYSIS_CHKPT"
+              "JGDAS_GLOBAL_OCEAN_ANALYSIS_POST")
+#              "JGDAS_GLOBAL_OCEAN_ANALYSIS_VRFY")
```
Comment (AndrewEichmann-NOAA) [inline test/soca/gw/CMakeLists.txt:37]: What's happening with VRFY?

Comment (guillaumevernieres) [reply test/soca/gw/CMakeLists.txt:37]: Waiting for your "post" PR to be merged in before we can run VRFY.

File: CMakeLists.txt [hunk 3]
```diff
@@ -87,6 +87,11 @@ if(BUILD_GDASBUNDLE)
   ecbuild_bundle( PROJECT fv3-jedi    GIT "https://github.com/jcsda/fv3-jedi.git"             BRANCH develop )
 
 # SOCA associated repositories
+# TODO: Move the Icepack fork to EMC github
+  set(BUILD_ICEPACK "ON" CACHE STRING "Build the icepack library")
```
Comment (CoryMartin-NOAA) [inline CMakeLists.txt:91]: brrr!

File: parm/soca/berror/soca_bump_split.yaml [hunk 4]
```diff
@@ -3,52 +3,55 @@ geometry:
...
-      - !ENV ${CVAR}
-    io:
-      files prefix: !ENV ${CVAR}
+      input model files:
+      - parameter: rh
+        file:
+          read_from_file: 1
+          date: *date
+          basename: ./
+          ocn_filename: ocn.cor_rh.incr.0001-01-01T00:00:00Z.nc
```
Comment (CoryMartin-NOAA) [inline parm/soca/berror/soca_bump_split.yaml:44]: we had ocean DA back in Jesus' time?

File: scripts/exgdas_global_marine_analysis_chkpt.sh [hunk 5]
```diff
@@ -0,0 +1,79 @@
+#!/bin/bash
```
Comment (CoryMartin-NOAA) [inline scripts/exgdas_global_marine_analysis_chkpt.sh]: note for later, g-w folks now want `#! /usr/bin/env bash` not `/bin/bash`

File: scripts/exgdas_global_marine_analysis_chkpt.sh [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+################################################################################
+####  UNIX Script Documentation Block
+#                      .                                             .
+# Script name:         exgdas_global_marine_analysis_chkpt.sh
+# Script description:  Prepare MOM6 IAU increment and insert the seaice analysis
+#                       in the CICE6 restart
+#
+# Author: Guillaume Vernieres        Org: NCEP/EMC     Date: 2023-04-01
+#
+# Abstract: Doodads for ocean and seaice DA
```
Comment (CoryMartin-NOAA) [inline scripts/exgdas_global_marine_analysis_chkpt.sh]: doodads!

File: scripts/exgdas_global_marine_analysis_prep.py [hunk 7]
```diff
@@ -223,7 +222,10 @@ def find_bkgerr(input_date, domain):
 
 # Variables of convenience
 half_assim_freq = timedelta(hours=int(os.getenv('assim_freq'))/2)
+assim_freq = timedelta(hours=int(os.getenv('assim_freq')))
```
Comment (CoryMartin-NOAA) [inline scripts/exgdas_global_marine_analysis_prep.py]: why? can you not make the next line `- assim_freq/2` and not even define `half_assim_freq`?

File: scripts/exgdas_global_marine_analysis_prep.py [hunk 8]
```diff
@@ -247,24 +249,27 @@ def find_bkgerr(input_date, domain):
...
 ################################################################################
 # stage static files
+
 logging.info(f"---------------- Stage static files")
 ufsda.stage.soca_fix(stage_cfg)
 
 ################################################################################
 # stage background error files
+
+logging.info(f"---------------- Stage static files")
```
Comment (CoryMartin-NOAA) [inline scripts/exgdas_global_marine_analysis_prep.py:258]: Sometime soon, should check out the workflow FileHandler for staging files

File: scripts/exgdas_global_marine_analysis_vrfy.py [hunk 9]
```diff
@@ -99,13 +117,16 @@ def plot_zonal_slice(config):
...
-    plt.savefig(figname, bbox_inches='tight')
+    plt.savefig(figname, bbox_inches='tight', dpi=600)
 
 
 comout = os.getenv('COMOUT')
 data = os.getenv('DATA')
+pdy = os.getenv('PDY')
 cyc = os.getenv('cyc')
-bcyc = str((int(cyc) - 3) % 24)
+bcyc = str((int(cyc) - 3) % 24).zfill(2)
```
Comment (CoryMartin-NOAA) [inline scripts/exgdas_global_marine_analysis_vrfy.py:128]: not really sure what the point of this is? if you want the previous cycle, why not use timedelta?

File: scripts/exgdas_global_marine_analysis_vrfy.py [hunk 10]
```diff
@@ -156,6 +177,81 @@ def plot_zonal_slice(config):
...
+    config.update({'variable': 'hsnon', 'bounds': [-0.1, 0.1], 'proj': proj})
+    plot_horizontal_slice(config)
+
+#######################################
+# Analysis/Background
+#######################################
+
+#######################################
+# Sea ice
+data_files = [os.path.join(comout, 'gdas.t'+cyc+'z.iceana.nc'),
```
Comment (CoryMartin-NOAA) [inline scripts/exgdas_global_marine_analysis_vrfy.py]: should not hard code `gdas` anywhere, that is `$RUN`. Presumably we will have an early cycle SOCA analysis right?

File: test/soca/CMakeLists.txt [hunk 11]
```diff
@@ -5,7 +5,9 @@ file(GLOB bkg_diag "${TESTDATA}/ocn_da_2018_04_15_*.nc")
 file(MAKE_DIRECTORY ${PROJECT_BINARY_DIR}/test/soca/bkg/RESTART)
 file(MAKE_DIRECTORY ${PROJECT_BINARY_DIR}/test/soca/bkg/RESTART_GES)
 file(COPY ${bkg_diag} DESTINATION ${PROJECT_BINARY_DIR}/test/soca/bkg/RESTART)
-file(COPY ${TESTDATA}/iced.2019-04-15-43200.nc DESTINATION ${PROJECT_BINARY_DIR}/test/soca/bkg/RESTART)
+configure_file(${TESTDATA}/iced.2019-04-15-43200.nc
```
Comment (CoryMartin-NOAA) [inline test/soca/CMakeLists.txt:8]: configure_file? why is this one special?

File: ush/socaincr2mom6.py [hunk 12]
```diff
@@ -42,8 +46,71 @@ def socaincr2mom6(incr, bkg, grid, incr_out):
...
+    ds_incr.to_netcdf(incr_out, mode='a')
+    return
+
+
+def trefincr2mom6(bkgfile, anlfile, ocngridfile):
+    """
+    Create tref increment from FV3 and interpolate it to MOM6 grid.
+
+    Parameters:
+    bkgfile (str): path to FV3 background file with tref (prob. sfcf006)
```
Comment (CoryMartin-NOAA) [inline ush/socaincr2mom6.py:82]: this is the Gaussian grid history file right?

File: ush/ufsda/yamltools.py [hunk 13]
```diff
@@ -1,9 +1,42 @@
...
 from ufsda.misc_utils import isTrue
+import pygw
+
+logging.basicConfig(format='%(asctime)s:%(levelname)s:%(message)s',
+                    level=logging.INFO, datefmt='%Y-%m-%d %H:%M:%S')
+
+
+def save_check(config, target, app='var'):
+    """
+    Remove obs spaces that point to non-existent file and save
```
Comment (CoryMartin-NOAA) [inline ush/ufsda/yamltools.py:16]: can't wait until we don't need a dumb hack like this...

**Review by CoryMartin-NOAA — COMMENTED** [review]
Main comment: don't hardcode `gdas` in places, use `RUN`, or at the very least some variable.

File: test/soca/gw/CMakeLists.txt (hunk 2 above)
Comment (ShastriPaturi) [reply test/soca/gw/CMakeLists.txt:37]: VRFY fails for me on hera even after running a python script with cartopy.

File: scripts/exgdas_global_marine_analysis_vrfy.py (hunk 9 above)
Comment (guillaumevernieres) [reply scripts/exgdas_global_marine_analysis_vrfy.py:128]: For timedelta, I vaguely remember not being happy with my clunky timedelta implementation. I'm fine with the low-level stuff above, unless you propose a better way that fits in 1 line ...

File: test/soca/CMakeLists.txt (hunk 11 above)
Comment (guillaumevernieres) [reply test/soca/CMakeLists.txt:8]: we use it as a warm restart but it's actually pretty cold?

...

I forgot ... checking right now.

File: scripts/exgdas_global_marine_analysis_vrfy.py (hunk 10 above)
Comment (guillaumevernieres) [reply scripts/exgdas_global_marine_analysis_vrfy.py]: I don't think vrfy will ever make it to nco, but will do.

File: ush/socaincr2mom6.py (hunk 12 above)
Comment (guillaumevernieres) [reply ush/socaincr2mom6.py:82]: yes

File: scripts/exgdas_global_marine_analysis_vrfy.py (hunk 10 above)
Comment (CoryMartin-NOAA) [reply scripts/exgdas_global_marine_analysis_vrfy.py]: ok well even so, we will want to compare gdas and gfs, no?

File: test/soca/CMakeLists.txt (hunk 11 above)
Comment (guillaumevernieres) [reply test/soca/CMakeLists.txt:8]: Ha! Because that file is renamed.

File: test/soca/gw/CMakeLists.txt (hunk 2 above)
Comment (guillaumevernieres) [reply test/soca/gw/CMakeLists.txt:37]: The bug fix will be in @AndrewEichmann-NOAA 's PR @ShastriPaturi .

**Comment by guillaumevernieres** [comment]
@CoryMartin-NOAA , I went through the marine DA scripts and replaced `cdump` and `gdas` with `RUN`, and a few other things.

**Review by ShastriPaturi — APPROVED** [review]
Works for me after the VRFY JJOB is commented out.

File: scripts/exgdas_global_marine_analysis_vrfy.py (hunk 10 above)
Comment (guillaumevernieres) [reply scripts/exgdas_global_marine_analysis_vrfy.py]: yes, for sure.

File: parm/soca/berror/soca_bump_split.yaml (hunk 4 above)
Comment (AndrewEichmann-NOAA) [reply parm/soca/berror/soca_bump_split.yaml:44]: Gotta do something with those reconstructions

**Review by AndrewEichmann-NOAA — APPROVED** [review]

**Review by CoryMartin-NOAA — APPROVED** [review]

**Pull request merged**